{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipywidgets as widgets\n",
    "from IPython.display import clear_output\n",
    "from intermine.webservice import Service\n",
    "from IPython.core.display import display, HTML\n",
    "import tabulate\n",
    "import json\n",
    "import re\n",
    "from ipywidgets import Button, GridBox, Layout, ButtonStyle\n",
    "import csv\n",
    "import pandas as pd\n",
    "\n",
    "## RUN ME FIRST\n",
    "import primer3.bindings as bindings\n",
    "\n",
    "primer__ = re.compile(\"PRIMER_((LEFT)|(RIGHT))_[0-5]\", re.I)\n",
    "entered_value = \"SMESG000027291.1\"\n",
    "\n",
    "def designPrimers(sequence):\n",
    "    PCR_primers = bindings.designPrimers(\n",
    "    {\n",
    "        'SEQUENCE_ID': 'MH1000',\n",
    "        'SEQUENCE_TEMPLATE': sequence,\n",
    "        'SEQUENCE_INCLUDED_REGION': [10,len(sequence)-10]\n",
    "    },\n",
    "    {\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_PICK_INTERNAL_OLIGO': 1,\n",
    "        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "        'PRIMER_MIN_SIZE': 18,\n",
    "        'PRIMER_MAX_SIZE': 25,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 57.0,\n",
    "        'PRIMER_MAX_TM': 63.0,\n",
    "        'PRIMER_MIN_GC': 20.0,\n",
    "        'PRIMER_MAX_GC': 80.0,\n",
    "        'PRIMER_MAX_POLY_X': 100,\n",
    "        'PRIMER_INTERNAL_MAX_POLY_X': 100,\n",
    "        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "        'PRIMER_DNA_CONC': 50.0,\n",
    "        'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "        'PRIMER_MAX_SELF_ANY': 12,\n",
    "        'PRIMER_MAX_SELF_END': 8,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "        'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "        'PRIMER_PRODUCT_SIZE_RANGE': [[750,1000],],\n",
    "    })\n",
    "\n",
    "    qPCR_primers = bindings.designPrimers(\n",
    "    {\n",
    "        'SEQUENCE_ID': 'MH1000',\n",
    "        'SEQUENCE_TEMPLATE': sequence,\n",
    "        'SEQUENCE_INCLUDED_REGION': [10,len(sequence)-10]\n",
    "    },\n",
    "    {\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_PICK_INTERNAL_OLIGO': 1,\n",
    "        'PRIMER_INTERNAL_MAX_SELF_END': 8,\n",
    "        'PRIMER_MIN_SIZE': 18,\n",
    "        'PRIMER_MAX_SIZE': 25,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 57.0,\n",
    "        'PRIMER_MAX_TM': 63.0,\n",
    "        'PRIMER_MIN_GC': 20.0,\n",
    "        'PRIMER_MAX_GC': 80.0,\n",
    "        'PRIMER_MAX_POLY_X': 100,\n",
    "        'PRIMER_INTERNAL_MAX_POLY_X': 100,\n",
    "        'PRIMER_SALT_MONOVALENT': 50.0,\n",
    "        'PRIMER_DNA_CONC': 50.0,\n",
    "        'PRIMER_MAX_NS_ACCEPTED': 0,\n",
    "        'PRIMER_MAX_SELF_ANY': 12,\n",
    "        'PRIMER_MAX_SELF_END': 8,\n",
    "        'PRIMER_PAIR_MAX_COMPL_ANY': 12,\n",
    "        'PRIMER_PAIR_MAX_COMPL_END': 8,\n",
    "        'PRIMER_PRODUCT_SIZE_RANGE': [[100,200],],\n",
    "    })\n",
    "    \n",
    "    \n",
    "    return PCR_primers, qPCR_primers\n",
    "\n",
    "\n",
    "def gene_output(search_query):\n",
    "    service = Service(\"http://planmine.mpi-cbg.de/planmine/service\")\n",
    "\n",
    "    query = service.new_query(\"Gene\")\n",
    "    query.add_view(\n",
    "    \"transcripts.sequence.residues\", \"transcripts.primaryIdentifier\",\n",
    "    \"primaryIdentifier\", \"symbol\")\n",
    "\n",
    "    query.add_constraint(\"Gene\", \"LOOKUP\", search_query, \"\", code = \"A\")\n",
    "\n",
    "    return query\n",
    "\n",
    "    \n",
    "\n",
    "def handle_primer_output(input_primer_dict):\n",
    "    primer_pairs = dict()\n",
    "    for x in input_primer_dict:\n",
    "        if primer__.search(x):\n",
    "            primer_pairs[x.split(\"_\")[2]] = {}   \n",
    "    \n",
    "    for x in primer_pairs:\n",
    "        for LoR in [\"LEFT\", \"RIGHT\"]: \n",
    "            penalty = input_primer_dict.get(f\"PRIMER_{LoR}_{x}_PENALTY\")\n",
    "            sequence = input_primer_dict.get(f\"PRIMER_{LoR}_{x}_SEQUENCE\")\n",
    "            boundaries = input_primer_dict.get(f\"PRIMER_{LoR}_{x}\")\n",
    "            melting_temp = input_primer_dict.get(f\"PRIMER_{LoR}_{x}_TM\")\n",
    "            GC_content = input_primer_dict.get(f\"PRIMER_{LoR}_{x}_GC_PERCENT\")\n",
    "            product_size = input_primer_dict.get(f'PRIMER_PAIR_{x}_PRODUCT_SIZE')\n",
    "            primer_size = input_primer_dict.get(f\"PRIMER_{LoR}_{x}_GC_PERCENT\")\n",
    "        \n",
    "            primer_pairs[x].update({LoR : {\"penalty\" : penalty, \n",
    "                                     \"sequence\" : sequence,\n",
    "                                     \"boundaries\" : boundaries,\n",
    "                                     \"TM\" : melting_temp, \n",
    "                                     \"GC\" : GC_content,\n",
    "                                          \"size\": boundaries[1],\n",
    "                                          \"product_size\" : product_size} })\n",
    "    return primer_pairs\n",
    "        \n",
    "def calcPCRout(sender):\n",
    "    global selected_transcript\n",
    "    sequence = all_entries.get(selected_transcript)\n",
    "    PCR_primers, qPCR_primers = designPrimers(sequence)\n",
    "    \n",
    "    clear_output()\n",
    "    \n",
    "    display(HTML(f'<h1>{selected_transcript}</h1>'))\n",
    "    display(HTML(f'<code>{sequence}</code>'))\n",
    "    \n",
    "    \n",
    "    PCR_primers_dict = handle_primer_output(PCR_primers)\n",
    "    qPCR_primers_dict = handle_primer_output(qPCR_primers)\n",
    "    \n",
    "    primer_dicts = {\"qPCR_primers\" : qPCR_primers_dict, \n",
    "                    \"PCR_primers\" : PCR_primers_dict}\n",
    "    \n",
    "   \n",
    "    \n",
    "    qPCR_out = widgets.Output()\n",
    "    PCR_out = widgets.Output()\n",
    "    \n",
    "\n",
    "    with qPCR_out:\n",
    "        for x, items in qPCR_primers_dict.items():\n",
    "            display(HTML(f\"<h3>Primer pair #{int(x)+1}\"))\n",
    "            \n",
    "            tables = list()\n",
    "            \n",
    "            for LoR in [\"LEFT\", \"RIGHT\"]:\n",
    "                boundaries = items.get(LoR).get(\"boundaries\")\n",
    "                if LoR == \"LEFT\":\n",
    "                    fr = \"Forward\"\n",
    "                elif LoR == \"RIGHT\":\n",
    "                    fr = \"Reverse\"\n",
    "                table = [[fr, \"PRIMER\"],\n",
    "                         [\"Sequence\", items.get(LoR).get(\"sequence\")],\n",
    "                     [\"From... To\", f\"{boundaries[0]}... {boundaries[0]+boundaries[1]}\"],\n",
    "                     [\"TM\", items.get(LoR).get(\"TM\")], \n",
    "                     [\"GC\", items.get(LoR).get(\"GC\")],\n",
    "                     [\"Amplicon Length\", items.get(LoR).get(\"product_size\")],\n",
    "                         [\"Primer Size\", items.get(LoR).get(\"size\")],\n",
    "                    ]\n",
    "                inter_out = widgets.Output()\n",
    "                with inter_out:\n",
    "                    display(HTML(tabulate.tabulate(table, tablefmt='html')))\n",
    "\n",
    "                tables.append(inter_out)\n",
    "        \n",
    "            griddy_2 = widgets.GridBox(children=tables,\n",
    "            layout=Layout(\n",
    "                width='100%',\n",
    "                grid_template_rows='auto',\n",
    "                grid_template_columns='50% 50%',))\n",
    "            display(griddy_2)\n",
    "            \n",
    "    with PCR_out: \n",
    "        for x, items in PCR_primers_dict.items():\n",
    "            display(HTML(f\"<h3>Primer pair #{int(x)+1}\"))\n",
    "            \n",
    "            tables = list()\n",
    "            \n",
    "            for LoR in [\"LEFT\", \"RIGHT\"]:\n",
    "                boundaries = items.get(LoR).get(\"boundaries\")\n",
    "                if LoR == \"LEFT\":\n",
    "                    fr = \"Forward\"\n",
    "                elif LoR == \"RIGHT\":\n",
    "                    fr = \"Reverse\"\n",
    "                table = [[fr, \"PRIMER\"],\n",
    "                         [\"Sequence\", items.get(LoR).get(\"sequence\")],\n",
    "                     [\"From... To\", f\"{boundaries[0]}... {boundaries[0]+boundaries[1]}\"],\n",
    "                     [\"TM\", items.get(LoR).get(\"TM\")], \n",
    "                     [\"GC\", items.get(LoR).get(\"GC\")],\n",
    "                     [\"Amplicon Length\", items.get(LoR).get(\"product_size\")],\n",
    "                         [\"Primer Size\", items.get(LoR).get(\"size\")],\n",
    "                    ]\n",
    "                inter_out = widgets.Output()\n",
    "                with inter_out:\n",
    "                    display(HTML(tabulate.tabulate(table, tablefmt='html')))\n",
    "\n",
    "                tables.append(inter_out)\n",
    "        \n",
    "            griddy_2 = widgets.GridBox(children=tables,\n",
    "                layout=Layout(\n",
    "                width='100%',\n",
    "                grid_template_rows='auto',\n",
    "                grid_template_columns='50% 50%',))\n",
    "            display(griddy_2)\n",
    "    \n",
    "    accordion = widgets.Accordion(children=[qPCR_out, PCR_out,])\n",
    "    accordion.set_title(0, 'qPCR Primers')\n",
    "    accordion.set_title(1, 'PCR Primers')\n",
    "    display(accordion)\n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "def create_calcPCR_butt():\n",
    "    global calcPCR, all_entries\n",
    "    calcPCR = widgets.Button(\n",
    "    description=\"CALC PCR PRIMERS\",\n",
    "    disabled=False,\n",
    "    button_style='info', # 'success', 'info', 'warning', 'danger' or ''\n",
    "    tooltip='Description',\n",
    "    icon='check'\n",
    "    )\n",
    "    display(calcPCR)\n",
    "    calcPCR.on_click(calcPCRout)\n",
    "    \n",
    "def select_transcript(widget):\n",
    "    global selected_transcript, tab\n",
    "    selected_transcript = (str(tab.children[tab.selected_index].children[0].value).strip(\"Name: \"))\n",
    "\n",
    "def printoutput(sender):\n",
    "    global entered_value, a, tab, all_entries, selected_transcript\n",
    "    search_query = str(a.value)\n",
    "    entered_value = search_query\n",
    "    clear_output()\n",
    "    all_entries = dict()\n",
    "    \n",
    "    if search_query.startswith(\"SMESG\"):\n",
    "        output = gene_output(search_query)\n",
    "        for x in list(output.rows()):\n",
    "            all_entries[x[\"transcripts.primaryIdentifier\"]] = x['transcripts.sequence.residues']\n",
    "        \n",
    "        tab_contents = [(f'Name: {x[\"transcripts.primaryIdentifier\"]}', \n",
    "                         f'Sequence: {x[\"transcripts.sequence.residues\"]}', \n",
    "                         f\"Length: {len(x['transcripts.sequence.residues'])}\") for x in list(output.rows())]\n",
    "\n",
    "        children = list()\n",
    "        \n",
    "        for entry in tab_contents:\n",
    "            items = [widgets.Label(str(entry[i])) for i in range(0,3)]\n",
    "            griddy = widgets.GridBox(items)\n",
    "            children.append(griddy)\n",
    "        \n",
    "        tab = widgets.Tab()\n",
    "        tab.children = children\n",
    "        for i in range(len(tab_contents)):\n",
    "            tab.set_title(i, tab_contents[i][0].strip(\"Name: \"))\n",
    "                         \n",
    "        display(tab)\n",
    "                         \n",
    "        selected_transcript = tab_contents[0][0].strip(\"Name: \")\n",
    "        \n",
    "        tab.observe(select_transcript, names='selected_index')\n",
    "                             \n",
    "        create_calcPCR_butt()\n",
    "            \n",
    "    else:\n",
    "        print(\"NOT A VALID FORMAT - ONLY SMESGs ARE ALLOWED.\")\n",
    "    \n",
    "   \n",
    "def start():\n",
    "    global entered_value, a\n",
    "    a = widgets.Text(value=entered_value, disabled=False)\n",
    "    display(a)\n",
    "    get_list = widgets.Button(\n",
    "    description=\"Get List\",\n",
    "    disabled=False,\n",
    "    button_style='info', # 'success', 'info', 'warning', 'danger' or ''\n",
    "    tooltip='Description',\n",
    "    icon='check'\n",
    "    )\n",
    "    display(get_list)\n",
    "\n",
    "    get_list.on_click(printoutput)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h1>SMEST027291002.1</h1>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<code>taactcatcttgttttctagatatgtagatggctcaatcataagggaaggaatttaattagttaagacatgagaataccgaactgcaccgggaattgttacaaatgagtcgggaaatgcagtcggtaaaagcaacattggtcaagctgacactggttaaacaagtatatcagatgctgacctaggctcaacgaggttctattatttatttaactcagaacctacagctatatgcccatcaacattatactcatttaaatatttaaaaattgtgtttagcaaagtaaattattcttttttttatttaaaattaattcatatgattttgtgtcaatatttccacaaaattttgtaaaatattttcaataaaataatttcaaagaaaatcataactgtattaacgaatatatcaaatacaaaaagatgatgtttgtttgaatattctttcttagcaaaattcctttgaattttaattttaagattatgctaatttgtcattgttcaatttgtcagcagtcttttagtgaaaattgtacagtatctgctttattatgtggacatgtttttcattcaaattgtcttacaatgtggctaaataacaagtcaaattgtccccaatgtagaagtccaaatgtcggaaattacattaaaaaattattttttgacaaaatgtctgataataataactcctttataaacgaaaataattcggttaacgaagaaaattttaagttgcggtcagaattagataaacttaaggatttaaactacaaaattacaaatgaattgaaaataaagtctgagtcaaatgagaatttgaaaaagaaactagatcagtctcaaagcactcaaatgtcacagtatttatcatttaccaaccaaatttatgaaacaaatgaaacaaatgagagatttaaaatgaaactagataaacttaaatttatgaatgatatcttaaaaaaatcaaaaccagaaatagatttatttattcaaaaattaccgaaagatgaacctccacaatttcaaatagaaaaattatgttctgcacttgtaacttgccaaaacgatttaaattattataagaaaaactctgatgaaatttccaagaaattagcaagaacaattttgaaaaattcaactcttaaaactgaaattgatcaattaaaagagcagatgagcaacaccgattcgaattttccggatttgaacacgagtgaacttctcgatactcccgttacagcgatgcgtacgaaaatgcggagacagttgacgactcctgaaatgctgttgtcttctccggatatgtcgaccaatcaacaaaatgaaaaacatttccgccagaaattactagaaaatccattcagatgcactccaaaacgatctctcagtcagaaaagctcttggaaattggacataatgattaatagatcgaaacccactcgattgccgcctccgcatattattcgactcgacagtgacacggaaaatcatccgatcgaggagccagttaggaaattgaaaaaagtacctcgttcgaaacacattaagccaatcaaaaaagattcaaccagttcgactccagtttatcattcgctctgaacgcgtttaatatgattcaatcgattgaaatatttaaaaattgatcgattttcttaccattgtaatggttaattaatattggttcgttgt</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "216e2a96257447cd9ba3e47c3dc61e4e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Accordion(children=(Output(), Output()), _titles={'0': 'qPCR Primers', '1': 'PCR Primers'})"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "start()                                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
